### PR TITLE
feature/JM-6482&6484: Monthly Utilisation Report

### PR DESCRIPTION
**Pull request by callum-r-young**
### JIRA link (if applicable) ###

https://centralgovernmentcgi.atlassian.net/browse/JM-6482
https://centralgovernmentcgi.atlassian.net/browse/JM-6484

### Change description ###

Added monthly utilisation report

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: client/templates/reporting/monthly-utilisation/prepare-select-month.njk [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+          id: "selectMonth",
+          name: "selectMonth",
+          label: {
+            text: "Select month to prepare report for",
+            classes: "govuk-label--m"
+          },
+          errorMessage: selectMonthError,
+          items: [
+            {
+              value: "01-2024",
```
Comment (gregs-cgi) [inline client/templates/reporting/monthly-utilisation/prepare-select-month.njk:39]: Probably need to offer more than just January!

File: server/routes/reporting/monthly-utilisation/monthly-utilisation.controller.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    return res.redirect(app.namedRoutes.build('reports.monthly-utilisation.prepare.confirm.get', {
+      month: req.body.selectMonth,
+    }));
+  };
+
+  module.exports.getCofirmPrepareMonthlyUtilisation = (app) => async(req, res) => {
+    const { month } = req.params;
+
+    try {
+      // GET FULL COURT NAME
```
Comment (gregs-cgi) [inline server/routes/reporting/monthly-utilisation/monthly-utilisation.controller.js:44]: !

File: server/routes/reporting/monthly-utilisation/monthly-utilisation.controller.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  };
+
+  module.exports.getFilterMonthlyUtilisation = (app) => (req, res) => {
+    const errors = req.session.errors || {};
+    const tmpBody = req.session.formFields || {};
+
+    delete req.session.errors;
+    delete req.session.formFields;
+
+    // HOW DO WE KNOW WHAT MONTHS TO DISPLAY
```
Comment (gregs-cgi) [inline server/routes/reporting/monthly-utilisation/monthly-utilisation.controller.js:65]: Comment - I guess we still need a list of available reports from the server?

File: server/routes/reporting/standard-report/definitions.js [hunk 4]
```diff
@@ -367,6 +368,29 @@ const { dailyUtilisationDAO, dailyUtilisationJurorsDAO } = require('../../../obj
...
+        headings: [
+          'dateFrom',
+          'reportDate',
+          'dateTo',
+          'reportTime',
+          '',
+          'courtName',
+        ],
+        bespokeReport: {
+          dao: async(req) => await monthlyUtilisationDAO.get(
```
Comment (ricardofreitasrocha) [inline server/routes/reporting/standard-report/definitions.js]: this needs updating to be `dao: (req) => monthlyUtilisationDAO.get()`

File: server/routes/reporting/standard-report/definitions.js [hunk 5]
```diff
@@ -367,6 +368,47 @@ const { dailyUtilisationDAO, dailyUtilisationJurorsDAO } = require('../../../obj
...
+        headings: [
+          'dateFrom',
+          'reportDate',
+          'dateTo',
+          'reportTime',
+          '',
+          'courtName',
+        ],
+        bespokeReport: {
+          dao: async(req) => await generateMonthlyUtilisationDAO.get(
```
Comment (ricardofreitasrocha) [inline server/routes/reporting/standard-report/definitions.js:382]: update to `dao: (req) => generateMonthlyUtilisationDAO.get(`

File: server/routes/reporting/standard-report/definitions.js [hunk 6]
```diff
@@ -367,6 +368,47 @@ const { dailyUtilisationDAO, dailyUtilisationJurorsDAO } = require('../../../obj
...
+      'view-monthly-utilisation': {
+        title: 'View monthly wastage and utilisation report',
+        headings: [
+          'courtName',
+          'reportDate',
+          '',
+          'reportTime',
+        ],
+        bespokeReport: {
+          dao: async(req) => await viewMonthlyUtilisationDAO.get(
```
Comment (ricardofreitasrocha) [inline server/routes/reporting/standard-report/definitions.js:401]: update to `dao: (req) => viewMonthlyUtilisationDAO.get(`.... reason is that the first call always fails

**Review by ricardofreitasrocha — CHANGES_REQUESTED** [review]

File: server/routes/reporting/standard-report/definitions.js (hunk 5 above)
Comment (callum-r-young) [reply server/routes/reporting/standard-report/definitions.js]: updated.

File: server/routes/reporting/standard-report/definitions.js (hunk 6 above)
Comment (callum-r-young) [reply server/routes/reporting/standard-report/definitions.js]: updated.

File: server/routes/reporting/monthly-utilisation/monthly-utilisation.controller.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+      return res.render('reporting/monthly-utilisation/confirm-prepare', {
+        courtName,
+        month: dateFilter(reportDate, 'yyyy-MM-DD', 'MMMM YYYY'),
+        incompleteServiceReportUrl: app.namedRoutes.build('reports.incomplete-service.filter.get'),
+        continueUrl: app.namedRoutes.build('reports.prepare-monthly-utilisation.report.get', { filter: reportDate }),
+        cancelUrl: app.namedRoutes.build('reports.statistics.get'),
+      });
+    } catch (e) {
+      console.log(e);
```
Comment (ricardofreitasrocha) [inline server/routes/reporting/monthly-utilisation/monthly-utilisation.controller.js:57]: could we handle this better? ie: add a logger call?

File: server/routes/reporting/monthly-utilisation/monthly-utilisation.controller.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        cancelUrl: app.namedRoutes.build('reports.statistics.get'),
+        tmpBody,
+        errors: {
+          title: 'Please check your search',
+          count: typeof errors !== 'undefined' ? Object.keys(errors).length : 0,
+          items: errors,
+        },
+      });
+    } catch (e) {
+      console.log(e);
```
Comment (ricardofreitasrocha) [inline server/routes/reporting/monthly-utilisation/monthly-utilisation.controller.js:90]: same here, add a logger call and handle better.

**Review by ricardofreitasrocha — CHANGES_REQUESTED** [review]

File: server/routes/reporting/monthly-utilisation/monthly-utilisation.controller.js (hunk 7 above)
Comment (callum-r-young) [reply server/routes/reporting/monthly-utilisation/monthly-utilisation.controller.js]: updated.

File: server/routes/reporting/monthly-utilisation/monthly-utilisation.controller.js (hunk 8 above)
Comment (callum-r-young) [reply server/routes/reporting/monthly-utilisation/monthly-utilisation.controller.js:99]: updated.

**Review by ricardofreitasrocha — APPROVED** [review]

**Pull request merged**